File: Leetcode/94-Binary-Tree-Inorder-Traversal/BinaryTreeInorderTraversal.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def inorderTraversal(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        res = []
        stack = []
        p = root
        while p is not None or len(stack)>0:
            while p is not None:
                stack.append(p)
                p = p.left
            if len(stack)>0:
                p = stack.pop()
                res.append(p.val)
                p = p.right
        return res

File: Leetcode/94-Binary-Tree-Inorder-Traversal/BinaryTreeInorderTraversal.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
var inorderTraversal = function(root) {
    var res = []
    var cur = root
    var prev
    while (cur) {
        if (!cur.left) {
            res.push(cur.val)
            cur = cur.right
            continue
        }
        prev = cur.left
        while (prev.right && prev.right!=cur) {
            prev = prev.right
        }
        if (!prev.right) {
            prev.right = cur
            cur = cur.left
        } else {
            res.push(cur.val)
            prev.right = null
            cur = cur.right
        }
    }
    return res
};
